### DISCUSSION TITLE
Ability to access middleware outside ctx (e.g. websockets middleware)

### DISCUSSION THREAD

**Discussion by backspacerhino** [adonisjs/core#1301] [Help]
After following the [official tutorial](https://preview.adonisjs.com/blog/socketio-with-adonisjs-v5/) for setting up Socket.io I have created a middleware function for websocket connection using `use()`

Inside my function I have token which is actually a generated token using [API Guard](https://preview.adonisjs.com/guides/auth/api-guard) from Adonis. I wish to check its validity but am unable to do so without ctx

```
...
myNamespace.use((socket, next) => {
    if (socket.handshake.query.token){
        const token = socket.handshake.query.token  // This is adonis generated token
        // How to check token validity 
        // If valid then next();
    }
    
});
```

One of the possible solutions would be to manually go through everything that OATGuard does when you checking for token validity
My question is if there is any other way of doing this and if not, could I ask for such an option?

**Comment by thetutlage** [comment]
What you really need here is to first class the [HttpContext](https://github.com/adonisjs/http-server/blob/develop/src/HttpContext/index.ts#L46) instance from the Websocket request details and then you will be able to use `ctx.auth.check()` directly.

**Reply by backspacerhino** [reply]
I don't think that's possible because the HttpContext requires `(request: RequestContract, response: ResponseContract, logger: LoggerContract, profiler: ProfilerRowContract)` and I can only provide request and maybe response

**Reply by thetutlage** [reply]
Logger and profiler are already part of the core and can be pulled easily from the container.  Here are docs on logger https://preview.adonisjs.com/guides/logger

```ts
import Logger from '@ioc:Adonis/Core/Logger'
import Profiler from '@ioc:Adonis/Core/Profiler'

new HttpContext(request, response, Logger, Profiler.create('websocket'))
```

**Reply by backspacerhino** [reply]
Do you have any recommendation on how to access `response`? Socket object doesnt have it. Also, I am not entirely sure that socket.request will for for request param but we will see

**Comment by sajaddp** [comment]
@backspacerhino Did you find a solution for use socket.io's middleware?

**Reply by backspacerhino** [reply]
At the end I "copied" the OATGuard logic so now I have something like this:

WebsocketMiddleware.ts
```
...
let savedToken: any
    try {
        savedToken = await (new WebsocketTokenAuth(repository)).run({ token: socket.handshake.query.token })
  }
  catch (error) {
       return error
  }
  
    //Join rooms
  const user = await new UserRepository().findOrFail(savedToken.userId)
  socket.join(user.id)
  next()
  ....
```

WebsocketTokenAuth.ts
```
...
export default class WebsocketTokenAuth {
  constructor (private repository: WebsocketAuthRepository) {
  }

  async run (data: { token: string | string[] | undefined }) {
    if (Array.isArray(data.token)) {
      data.token = data.token.join("")
    }

    if (!data.token) {
      throw new Error('MISSING_AUTHORIZATION_HEADER')
    }

    const explodedToken = data.token.split('.')
    const tokenId = Buffer.from(explodedToken[0], 'base64').toString('utf-8')
    let hashedToken: string

    try {
      hashedToken = createHash('sha256').update(explodedToken[1]).digest('hex')
    }
    catch (error) {
      throw new Error('INVALID_TOKEN')
    }

    const savedToken = await this.repository.findOrFail(tokenId)

    if (hashedToken !== savedToken.token) {
      throw new Error('INVALID_TOKEN')
    }

    // Check if token expired         
    if (savedToken.expiresAt && DateTime.local() > savedToken.expiresAt) {
      throw new Error('INVALID_TOKEN')
    }

    return savedToken
  }
}

```
I hope this helps
